utils/tra_utils.py

```diff
@@ -0,0 +1,213 @@
+        all_idents_seq.sort()
+        return all_idents_seq
+
+    def get_tra_file_paths(self) -> Sequence[Path]:
+        root_path = self.root_path
+        tra_dir = root_path / 'translations'
+        if not tra_dir.exists():
+            raise RuntimeError(f"Translation directory doesn't exist for file: '{self.name}' at root '{self.root_path}'")
+        if not tra_dir.is_dir():
+            raise RuntimeError(f"'translations' isn't a directory for file: '{self.name}' at root '{self.root_path}'")
+
+        lang_dirs = [child for child in tra_dir.iterdir() if child.is_dir()]
+        tra_name = self.name.replace('.d', '.tra')
+        tra_files_found = []
+        tra_files_missing = []
+        for lang_dir in lang_dirs:
+            tra_file = lang_dir / tra_name
+            if tra_file.exists():
+                tra_files_found.append(tra_file)
+            else:
+                tra_files_missing.append(tra_file)
+
+        if tra_files_missing:
+            tra_files_missing_str = ''.join([
+                f'\n * {fname}' for fname in tra_files_missing
+            ])
+            raise RuntimeError(f"Missing translation files:{tra_files_missing_str}")
+
+        return tra_files_found
+
+    def check_string_identifiers(self) -> None:
+        idents = set(self.gather_string_identifiers())
+
+        # if the file contains no string identifiers,
+        # it doesn't require .tra files, as they would be empty anyway
+        if not idents:
+            return
+
+        tra_files = self.get_tra_file_paths()
+        for tra_file_path in tra_files:
+            tra_dir = tra_file_path.parent.name
+            with TraFile.from_path(tra_file_path) as tra_file:
+                ident_strings = tra_file.gather_strings()
+            tra_idents = set(ident_strings.keys())
+            missing_idents = idents - tra_idents
+            if missing_idents:
+                missing_files_str = ''.join([
+                    f'\n * {ident}' for ident in missing_idents
+                ])
+                raise RuntimeError(f"Missing identifiers for file '{self.name}' in '{tra_dir}/{tra_file_path.name}': {missing_files_str}")
+
+
+def _get_whole_string(io: TextIO, line: str) -> str:
+    string_parts = []
+    where_end = line.find('~')
+    if where_end != -1:
+        string_parts.append(line[:where_end])
+    else:
+        string_parts.append(line)
+        while True: # do-until
+            line = io.readline()
+            if line == '':
+                break
+
+            where_end = line.find('~')
+            if where_end != -1:
+                new_part = line[:where_end]
+                string_parts.append(new_part)
+                break
+            else:
+                string_parts.append(line)
+
+    string = ''.join(string_parts)
+    return string
+
+def _get_ident_string(io: TextIO, line: Optional[str] = None) -> Optional[Tuple[str, str]]:
+    pat = re.compile('^(@[0-9]*) *= *~')
```

Comment:
i'd move this regex to some variable/const

Reply:
This original regex was quite situational, but after the discussion, it turned out there should be no problem with reading the whole file and using a multi-line regex. Thanks to that, the new regex is more stand-alone and could be factored out just fine.